enhancements/local-storage/automatic-detection.md

```diff
@@ -0,0 +1,157 @@
+
+### Goals
+
+- Automatically detection of available disks from nodes which can be used as PVs for OpenShift-cluster.
+- Should respond to the attach/detach events of the disks/devices.
+- Should have options for excluding a particular kind of disks.
+
+### Non-Goals
+
+## Proposal
+
+`local-storage-operator`is already capable of consuming local disks from the nodes and provisioning PVs out of it, but the disk/device paths needs to explicitly specified in the `LocalVolume` CR. The idea is to bring in one more layer of CR above it called `AutoDetectVolumes`.
+The controller for this CR will be responsible for automatically detecting disks/devices from the available nodes and creating/managing the life cycle of `LocalVolumes` which are governed by the `AutoDetectVolumes`.s
+
+### Risks and Mitigations
+
+N/A
+
+## Design Details
+
+API scheme for `AutoDetectVolumes`:
+
+```
+type DeviceDiscoveryPolicyType string
+
+const (
+	SSDs     DeviceDiscoveryPolicyType = "ssd"
+	Disks    DeviceDiscoveryPolicyType = "disk"
+	Partions DeviceDiscoveryPolicyType = "partitions"
+)
+
+type AutoDetectVolumeList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []AutoDetectVolume `json:"items"`
+}
+
+type AutoDetectVolume struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              AutoDetectVolumeSpec   `json:"spec"`
+	Status            AutoDetectVolumeStatus `json:"status,omitempty"`
+}
+
+type AutoDetectVolumeSpec struct {
+	// StorageClass name to use for set of matched devices
+	StorageClassName string `json:"storageClassName"`
+	// Volume mode. Raw or with file system
+	// + optional
+	VolumeMode PersistentVolumeMode `json:"volumeMode,omitempty"`
+	// File system type
+	// +optional
+	FSType string `json:"fsType,omitempty"`
+	// Nodes on which the autoDetection must run
+	// +optional
+	NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`
+	// DeviceDiscoverPolicies are the list of policies that AutoDetectVolume's controller
+	// will be looking for during detection
+	DeviceDiscoveryPolicies []DeviceDiscoveryPolicy `json:"deviceDiscoveryPolicies,omitempty"`
+}
+
+type AutoDetectVolumeStatus struct {
+	Status string `json:"status"`
+}
+
+type DeviceDiscoveryPolicy struct {
+	// Device type that should be used for auto detection. This would be one of the types supported
+	// by the local-storage operator. Initial possible types can be - crypt,raid1, raid4, raid5,
+	// raid10, multipath, disk, tape, printer, processor, worm, rom, scanner, mo-disk, changer,
+	// comm, raid, enclosure, rbc, osd, and no-lun.
+	DeviceType DeviceDiscoveryPolicyType `json:"deviceType"`
+	// A list of regular expressions that will be used to exclude certain devices
+	// For example - ["^rbd[0-9]+p?[0-9]{0,}$"]
+	// +optional
+	DeviceExclusionFilter []string `json:"deviceExclusionFilter"`
```

Comment:
How are these elements combined (AND or OR)--- Must it match all to be excluded or just one?

Should there be a DeviceInclusionFilter (that defaults to `".*"`)?

Reply:
its a OR combination. If anything matches it will be excluded. There will be a default Exclusion Filter for sure.

Reply:
(isType1 OR isType2 ...  OR isTypeN ) AND (!matchesFilter1 OR !matchesFilter2 ... OR !matchesFilter3)

right?

Reply:
I would suggest starting with an inclusion filter. By default, exclusion could just be devices that we detect as being in use (for example, have a filesystem). As a bonus, the exclusion filter could also remove devices from the list. I'd also call it `DeviceNameInclusionFilter` so it's obviously filtering on the name.

Reply:
+1 for inclusion pattern rather than exclusion

Reply:
added the inclusion pattern